In [1]:
import numpy as np
import random

# Question 1: Single Perceptron



In [2]:
# --- Dataset ---

# "1" patterns
patterns_1 = np.array([
    [0,0,1,0,0, 0,0,1,0,0, 0,0,1,0,0, 0,0,1,0,0, 0,0,1,0,0],
    [0,1,1,0,0, 0,0,1,0,0, 0,0,1,0,0, 0,0,1,0,0, 0,1,1,1,0],
    [0,0,1,0,0, 0,1,1,0,0, 0,0,1,0,0, 0,0,1,0,0, 0,0,1,0,0],
    [0,0,0,1,0, 0,0,1,1,0, 0,0,1,0,0, 0,1,1,0,0, 0,1,0,0,0],
    [0,0,0,0,1, 0,0,0,1,0, 0,0,1,0,0, 0,1,0,0,0, 1,0,0,0,0],
    [0,0,0,0,1, 0,0,1,0,0, 0,0,1,0,0, 0,0,1,0,0, 0,0,1,0,0]
])

# "0" patterns
patterns_0 = np.array([
    [0,1,1,1,0, 0,1,0,1,0, 0,1,0,1,0, 0,1,0,1,0, 0,1,1,1,0],
    [0,0,1,0,0, 0,1,0,1,0, 0,1,0,1,0, 0,1,0,1,0, 0,0,1,0,0],
    [0,0,1,0,0, 0,1,0,1,0, 0,1,0,1,0, 0,1,0,1,0, 0,1,1,1,0],
    [0,0,1,1,0, 0,1,0,0,1, 0,1,0,0,1, 0,1,0,1,0, 0,1,1,0,0],
    [1,1,1,1,1, 1,0,0,0,1, 1,0,1,0,1, 1,0,0,0,1, 1,1,1,1,1],
    [0,1,1,1,0, 1,0,0,0,1, 1,0,0,0,1, 1,0,0,0,1, 0,1,1,1,0]
])

In [3]:
# --- Prepare Training and Test Sets ---

# Create a list of indices [0, 1, 2, 3, 4, 5]
indices = list(range(6))
random.shuffle(indices)

# Choose 4 for training, 2 for testing
train_indices = indices[:4]
test_indices = indices[4:]

# Create the training set
train_1 = patterns_1[train_indices]
train_0 = patterns_0[train_indices]

# Create the test set
test_1 = patterns_1[test_indices]
test_0 = patterns_0[test_indices]

# Combine patterns and create labels
# Training data with labels (+1 for "1", -1 for "0")
training_data = list(zip(np.concatenate((train_1, train_0)), np.array([1]*4 + [-1]*4)))

# Test data with labels
test_data = list(zip(np.concatenate((test_1, test_0)), np.array([1]*2 + [-1]*2)))

# Shuffle the training data for random presentation order 
random.shuffle(training_data)

print(f"Total training patterns: {len(training_data)}")
print(f"Total test patterns: {len(test_data)}")

Total training patterns: 8
Total test patterns: 4


In [4]:
# --- Perceptron Training ---

# Step 1: Initialize parameters
learning_rate = 0.1
# Weights: 25 small random numbers
weights = np.random.rand(25) * 0.1 
# Bias: one small random number
bias = np.random.rand(1) * 0.1

# Epoch tracking
epochs = 0

In [5]:
while True:
    epochs += 1
    total_error = 0
    
    # Step 2: Shuffle the training data for each epoch
    random.shuffle(training_data)
    
    # Loop through each training pattern and its label
    for x, d in training_data:
        # Step 3: Calculate the network output for this input 'x'
        
        # 1. Calculate the raw output (dot product of x and weights, plus bias)
        raw_output = np.dot(x, weights) + bias

        # 2. Apply the sign activation function. If the result is >= 0, output is 1, else -1.
        if raw_output >= 0:
            y = 1
        else:
            y = -1

        # Step 4: Update the weights and bias if there's an error
        if y != d:
            total_error += 1

            # 3. Calculate the error term: (d - y)
            error = d - y
            # 4. Update the weights vector using the formula
            weights += learning_rate * error * x
            # 5. Update the bias using the formula
            bias += learning_rate * error

    # Step 5: Check the stopping condition
    # If total_error is 0 after a full epoch, the training is complete.
    if total_error == 0:
        print(f"Training complete after {epochs} epochs.")
        break
        
    # Optional: Print progress
    print(f"Epoch {epochs}, Total Errors: {total_error}")

# After the loop, print the final learned parameters
print("\nFinal Learned Weights:")
print(weights)
print("\nFinal Learned Bias:")
print(bias)

Epoch 1, Total Errors: 2
Epoch 2, Total Errors: 2
Training complete after 3 epochs.

Final Learned Weights:
[-0.19514708 -0.14023342 -0.15751726 -0.18349817  0.03378337 -0.12237197
 -0.14367376  0.21721796  0.08027355 -0.12371434 -0.10231235 -0.15847014
  0.21080518 -0.18937152 -0.17646172 -0.19789657  0.09088766  0.26243505
 -0.10271157 -0.17928206  0.06531002 -0.38947393 -0.15415724 -0.36419489
 -0.15171093]

Final Learned Bias:
[0.02215597]


Results from running the perceptron once:

Epoch 1, Total Errors: 2  
Epoch 2, Total Errors: 2  
Training complete after 3 epochs.

**Final Learned Weights:**
```python
[-0.19514708, -0.14023342, -0.15751726, -0.18349817,  0.03378337, -0.12237197,
 -0.14367376,  0.21721796,  0.08027355, -0.12371434, -0.10231235, -0.15847014,
  0.21080518, -0.18937152, -0.17646172, -0.19789657,  0.09088766,  0.26243505,
 -0.10271157, -0.17928206,  0.06531002, -0.38947393, -0.15415724, -0.36419489,
 -0.15171093]

Final Learned Bias:
[0.02215597]


In [6]:
# --- Evaluate the Perceptron on the Test Set ---

test_errors = 0
total_test_patterns = len(test_data)

# Loop through each test pattern and its true label
for x, d in test_data:
    # Use the FINAL trained weights and bias to make a prediction
    
    # 1. Calculate the raw output
    raw_output = np.dot(x, weights) + bias
    
    # 2. Apply the sign activation function
    if raw_output >= 0:
        y = 1
    else:
        y = -1
    
    # 3. Check if the prediction is incorrect
    if y != d:
        test_errors += 1

# Calculate and print the performance
accuracy = (total_test_patterns - test_errors) / total_test_patterns
error_rate = test_errors / total_test_patterns

print(f"Test Results:")
print(f"Correct classifications: {total_test_patterns - test_errors} / {total_test_patterns}")
print(f"Accuracy: {accuracy * 100:.2f}%")
print(f"Error Rate: {error_rate * 100:.2f}%")

Test Results:
Correct classifications: 3 / 4
Accuracy: 75.00%
Error Rate: 25.00%


**Results from running the perceptron once:**

**Test Results:**
- Correct classifications: `3 / 4`
- Accuracy: `75.00%`
- Error Rate: `25.00%`

In [7]:
# --- Run Multiple Experiments for Average Performance ---

num_experiments = 10
accuracies = []

for i in range(num_experiments):
    print(f"--- Running Experiment {i+1}/{num_experiments} ---")
    
    # 1. Split the data randomly
    indices = list(range(6))
    random.shuffle(indices)
    train_indices = indices[:4]
    test_indices = indices[4:]

    training_data = list(zip(np.concatenate((patterns_1[train_indices], patterns_0[train_indices])), np.array([1]*4 + [-1]*4)))
    test_data = list(zip(np.concatenate((patterns_1[test_indices], patterns_0[test_indices])), np.array([1]*2 + [-1]*2)))

    
    # 2. Initialize parameters
    weights = np.random.rand(25) * 0.1
    bias = np.random.rand(1) * 0.1
    epochs = 0

    # 3. Train the model
    while True:

        total_error = 0
        for x, d in training_data:
            # ...
            if (np.dot(x, weights) + bias) >= 0: y=1
            else: y=-1
            
            if y != d:
                total_error += 1
                error = d - y
                weights += learning_rate * error * x
                bias += learning_rate * error
        
        if total_error == 0:
            break
            
    # 4. Evaluate on the test set
    test_errors = 0

    for x, d in test_data:
        if (np.dot(x, weights) + bias) >= 0: y=1
        else: y=-1
        if y != d: test_errors += 1

    accuracy = (len(test_data) - test_errors) / len(test_data)
    accuracies.append(accuracy)
    print(f"Experiment {i+1} Accuracy: {accuracy*100:.2f}%")

# 5. Calculate average performance
average_accuracy = np.mean(accuracies)
print(f"\n--- Final Results ---")
print(f"Average accuracy over {num_experiments} experiments: {average_accuracy * 100:.2f}%")

--- Running Experiment 1/10 ---
Experiment 1 Accuracy: 100.00%
--- Running Experiment 2/10 ---
Experiment 2 Accuracy: 100.00%
--- Running Experiment 3/10 ---
Experiment 3 Accuracy: 100.00%
--- Running Experiment 4/10 ---
Experiment 4 Accuracy: 75.00%
--- Running Experiment 5/10 ---
Experiment 5 Accuracy: 75.00%
--- Running Experiment 6/10 ---
Experiment 6 Accuracy: 75.00%
--- Running Experiment 7/10 ---
Experiment 7 Accuracy: 100.00%
--- Running Experiment 8/10 ---
Experiment 8 Accuracy: 75.00%
--- Running Experiment 9/10 ---
Experiment 9 Accuracy: 100.00%
--- Running Experiment 10/10 ---
Experiment 10 Accuracy: 75.00%

--- Final Results ---
Average accuracy over 10 experiments: 87.50%



## Experiment: Run perceptron 10 times, each with a different random shuffle of training and test data, and random initial weight and bias 

### Running Experiments
- **Experiment 1**: Accuracy: **100.00%**
- **Experiment 2**: Accuracy: **100.00%**
- **Experiment 3**: Accuracy: **100.00%**
- **Experiment 4**: Accuracy: **75.00%**
- **Experiment 5**: Accuracy: **75.00%**
- **Experiment 6**: Accuracy: **75.00%**
- **Experiment 7**: Accuracy: **100.00%**
- **Experiment 8**: Accuracy: **75.00%**
- **Experiment 9**: Accuracy: **100.00%**
- **Experiment 10**: Accuracy: **75.00%**

---

## Final Results
**Average accuracy over 10 experiments**: **87.50%**
